### PR TITLE
Feature: Optimized memory management on DSP

### PR DESCRIPTION
**Pull request by Critsium-xy**
Fully make use of the memory of dsp hardware. Can run small cases now but still under testing. DSP memory management is very weird so maybe more time is needed on this.

### CODE REVIEW

**Comment by Critsium-xy** [comment]
Currently this can successfully run on small cases but will cause immediate memory failure on large cases.

**Review by mohanchen — APPROVED** [review]

File: source/module_hsolver/diago_dav_subspace.cpp [hunk 1]
```diff
@@ -445,6 +445,8 @@ void Diago_DavSubspace<T, Device>::cal_elem(const int& dim,
     if (this->diag_comm.nproc > 1)
     {
         auto* swap = new T[notconv * this->nbase_x];
+        auto* target = new T[notconv * this->nbase_x];
```
Comment (mohanchen) [inline source/module_hsolver/diago_dav_subspace.cpp]: update again to not affecting the original code?

Comment (Critsium-xy) [reply source/module_hsolver/diago_dav_subspace.cpp]: Updated this in the latest commit and added some code. It is ok to merge after getting the tests passed.

File: source/module_hsolver/diago_dav_subspace.cpp [hunk 2]
```diff
@@ -444,7 +444,72 @@ void Diago_DavSubspace<T, Device>::cal_elem(const int& dim,
 #ifdef __MPI
     if (this->diag_comm.nproc > 1)
     {
+#ifdef __DSP
```
Comment (mohanchen) [inline source/module_hsolver/diago_dav_subspace.cpp:448]: I would suggest DSP codes can be organized into another source file

**Review by mohanchen — APPROVED** [review]

**Pull request merged**